File: exposed/src/main/kotlin/dev/cerbos/queryplan/exposed/AttributeMapping.kt
package dev.cerbos.queryplan.exposed

import org.jetbrains.exposed.v1.core.Alias
import org.jetbrains.exposed.v1.core.Column
import org.jetbrains.exposed.v1.core.Op
import org.jetbrains.exposed.v1.core.Table

/**
 * Resolves a plan's attribute reference, e.g. `request.resource.attr.owner`, onto storage.
 *
 * Resolution is fail-closed: returning `null` makes the translation raise
 * [UnmappedAttributeException] rather than guess a column name. [cerbosMapping] builds the usual
 * static table; implement this interface directly to resolve references by rule.
 *
 * A dotted reference is matched against the LONGEST mapped prefix, and the remainder is walked
 * through that relation's nested [AttributeMapping.Relation.fields]. A function-style resolver is
 * therefore asked for progressively shorter prefixes of the same reference.
 */
public fun interface AttributeResolver {
    public fun resolve(reference: String): AttributeMapping?
}

/** What an attribute reference resolves to: a column, or a relation reaching another table. */
public sealed interface AttributeMapping {

    /** One column on the table of the scope that declares it. */
    public class Field internal constructor(
        public val column: Column<*>,
        /**
         * This attribute's NULL convention, or `null` when nothing is declared. Declaring it
         * asserts two things at once: that the column can be NULL, and how that NULL reaches
         * `check()`.
         *
         * - [NullAttributeRepresentation.EXPLICIT]: the NULL is sent as a null VALUE, so the
         *   equality family (`eq`, `ne`, `in`) renders DEFINITELY and a negation includes the NULL
         *   rows CEL allows.
         * - [NullAttributeRepresentation.OMITTED]: the NULL sends no attribute, so a null operand
         *   against this attribute is refused whatever the call-level option says.
         * - `null`: the column renders the way an undeclared column always has, as if NOT NULL, and
         *   [Options.nullAttributeRepresentation] decides only whether a null OPERAND is refused.
         *
         * `null` does NOT mean "inherit the call-level option" for rendering, and must not: the
         * call-level default is EXPLICIT, so inheriting it would hand definite equality to every
         * undeclared column and return NULL rows the PDP denies for an attribute the caller omits.
         * The fix is opt-in per column (docs/adr/0004).
         */
        public val nullAttributeRepresentation: NullAttributeRepresentation?,
    ) : AttributeMapping {
        override fun toString(): String =
            "Field(${column.table.tableName}.${column.name}, nulls=$nullAttributeRepresentation)"
    }

    /**
     * A relation to another table, lowered to a correlated subquery over that table read BARE.
     *
     * The correlation is always `<table>.<to> = <parent>.<from>`, where the parent is the scope
     * that declares the relation: the root table at the top level, the enclosing relation's table
     * when nested under [fields]. Both keys are single columns, so a composite association key is
     * not expressible here: that is a compile error rather than a wrong join.
     */
    public class Relation internal constructor(
        public val cardinality: Cardinality,
        /** The related table: the one a lambda body, or a to-one path, reads its fields from. */
        public val table: Table,
        /** The column on the PARENT scope's table. */
        public val from: Column<*>,
        /** The matching column on [table]. */
        public val to: Column<*>,
        /**
         * The column that stands for the element itself: the value of a scalar collection such as
         * `tagNames`, or the default member when an object collection is used as a bare value.
         */
        public val element: Field?,
        /** The element's own attributes. An entry may itself be a [Relation], giving a chain. */
        public val fields: Map<String, AttributeMapping>,
        /**
         * The predicate the APPLICATION applies whenever it reads [table]: a soft-delete flag, a
         * tenant column, a subtype discriminator. The subquery reads the table bare and cannot see
         * any of those, so without this it examines rows the application never serialised into the
         * resource attributes. It is built against the alias each subquery instance reads through.
         */
        internal val visibleWhen: ((Alias<Table>) -> Op<Boolean>)?,
    ) : AttributeMapping {

        public enum class Cardinality {
            /**
             * At most one related row. Read as a scalar; an absent row is a missing attribute.
             * Nothing makes the database enforce the single row: [to] needs a unique index.
             */
            ONE,

            /** Any number of related rows: a collection. */
            MANY,
        }

        /** Whether the caller declared a [visibleWhen] predicate. */
        public val hasVisibilityPredicate: Boolean get() = visibleWhen != null

        init {
            require(to.table === table) {
                "Relation target column ${to.table.tableName}.${to.name} is not a column of ${table.tableName}"
            }
            element?.let {
                require(it.column.table === table) {
                    "Relation element column ${it.column.table.tableName}.${it.column.name} " +
                        "is not a column of ${table.tableName}"
                }
            }
            fields.forEach { (name, mapping) ->
                val owner = when (mapping) {
                    is Field -> mapping.column.table
                    is Relation -> mapping.from.table
                }
                require(owner === table) {
                    "Relation field '$name' reads ${owner.tableName}, not the relation's table ${table.tableName}"
                }
            }
        }

        override fun toString(): String =
            "Relation($cardinality ${table.tableName}: ${to.name} = ${from.table.tableName}.${from.name})"
    }

    public companion object {
        /**
         * A column that declares NO null convention, and so renders as if it were NOT NULL. See
         * [Field.nullAttributeRepresentation]: the call-level [Options.nullAttributeRepresentation]
         * decides only whether a null OPERAND against it is refused, never how it renders.
         */
        @JvmStatic
        public fun field(column: Column<*>): Field = Field(column, null)

        /** A column that declares its own NULL convention. */
        @JvmStatic
        public fun field(column: Column<*>, nulls: NullAttributeRepresentation): Field = Field(column, nulls)
    }
}

/** A static reference-to-mapping table, as built by [cerbosMapping]. */
public class AttributeMappings internal constructor(entries: Map<String, AttributeMapping>) : AttributeResolver {
    /** The mapped references. A defensive copy: later changes to the source map never reach it. */
    public val entries: Map<String, AttributeMapping> = LinkedHashMap(entries)

    override fun resolve(reference: String): AttributeMapping? = entries[reference]

    override fun toString(): String = "AttributeMappings(${entries.keys})"

    public companion object {
        /** For callers that assemble the table without the Kotlin DSL. */
        @JvmStatic
        public fun of(entries: Map<String, AttributeMapping>): AttributeMappings = AttributeMappings(entries)
    }
}

@DslMarker
public annotation class CerbosMappingDsl

/**
 * Builds a static mapping:
 *
 * ```kotlin
 * val mapping = cerbosMapping {
 *     "request.resource.id" to Documents.id
 *     "request.resource.attr.owner" to field(Documents.owner, nulls = NullAttributeRepresentation.EXPLICIT)
 *     "request.resource.attr.tags" to many(Tags, from = Documents.id, to = Tags.documentId) {
 *         "name" to Tags.name
 *         visibleWhen { t -> t[Tags.deletedAt].isNull() }
 *     }
 * }
 * ```
 */
public fun cerbosMapping(block: MappingBuilder.() -> Unit): AttributeMappings =
    AttributeMappings(MappingBuilder().apply(block).entries)

/** Shared by the top-level mapping and every nested relation: the entries and their factories. */
@CerbosMappingDsl
public open class MappingBuilder internal constructor() {
    internal val entries: MutableMap<String, AttributeMapping> = LinkedHashMap()

    /**
     * Maps [this] reference onto a column that declares no null convention, and so renders as if
     * it were NOT NULL ([AttributeMapping.Field.nullAttributeRepresentation]).
     */
    public infix fun String.to(column: Column<*>) {
        put(this, AttributeMapping.Field(column, null))
    }

    /** Maps [this] reference onto a [field], [one] or [many] mapping. */
    public infix fun String.to(mapping: AttributeMapping) {
        put(this, mapping)
    }

    /** A column that declares its own NULL convention. */
    public fun field(column: Column<*>, nulls: NullAttributeRepresentation? = null): AttributeMapping.Field =
        AttributeMapping.Field(column, nulls)

    /** A to-one relation: `table.to = parent.from`, at most one row. */
    public fun <T : Table> one(
        table: T,
        from: Column<*>,
        to: Column<*>,
        block: RelationBuilder<T>.() -> Unit = {},
    ): AttributeMapping.Relation =
        RelationBuilder<T>().apply(block).build(AttributeMapping.Relation.Cardinality.ONE, table, from, to, null)

    /**
     * A to-many relation: `table.to = parent.from`. Pass [element] for a scalar collection, where
     * each element IS a value (`tagNames`), or as the default member of an object collection.
     */
    public fun <T : Table> many(
        table: T,
        from: Column<*>,
        to: Column<*>,
        element: Column<*>? = null,
        block: RelationBuilder<T>.() -> Unit = {},
    ): AttributeMapping.Relation =
        RelationBuilder<T>().apply(block).build(AttributeMapping.Relation.Cardinality.MANY, table, from, to, element)

    private fun put(reference: String, mapping: AttributeMapping) {
        require(reference.isNotEmpty()) { "An attribute reference must not be empty" }
        require(entries.put(reference, mapping) == null) { "'$reference' is mapped twice" }
    }
}

/** The body of a [MappingBuilder.one] or [MappingBuilder.many] block. */
@CerbosMappingDsl
public class RelationBuilder<T : Table> internal constructor() : MappingBuilder() {
    private var visibleWhen: ((Alias<T>) -> Op<Boolean>)? = null

    /**
     * Declares the predicate the application itself applies when it reads this relation's table.
     * It receives the alias the subquery reads through: resolve columns with `alias[Table.column]`.
     * It is ANDed into the subquery's correlation, so it narrows the rows the subquery EXAMINES,
     * which is what keeps it right under negation.
     */
    public fun visibleWhen(predicate: (Alias<T>) -> Op<Boolean>) {
        require(visibleWhen == null) { "visibleWhen is declared twice" }
        visibleWhen = predicate
    }

    internal fun build(
        cardinality: AttributeMapping.Relation.Cardinality,
        table: T,
        from: Column<*>,
        to: Column<*>,
        element: Column<*>?,
    ): AttributeMapping.Relation {
        val predicate = visibleWhen
        return AttributeMapping.Relation(
            cardinality = cardinality,
            table = table,
            from = from,
            to = to,
            element = element?.let { AttributeMapping.Field(it, null) },
            fields = LinkedHashMap(entries),
            // The alias a subquery hands in is always an alias of `table`, so the narrowing is safe;
            // the type parameter only exists to give the caller `alias[Table.column]` with no cast.
            visibleWhen = predicate?.let { typed ->
                { alias -> @Suppress("UNCHECKED_CAST") typed(alias as Alias<T>) }
            },
        )
    }
}

File: sequel/lib/cerbos/sequel/attribute_mapping.rb
# frozen_string_literal: true

require_relative "errors"

module Cerbos
  module Sequel
    # Maps one Cerbos attribute reference to the Sequel model. The attribute reference is the
    # +variable+ name in a query plan, for example <tt>"request.resource.attr.ownerId"</tt>.
    # The mapping is a scalar {Field} or a collection {Association}.
    #
    #   MAPPING = {
    #     "request.resource.attr.ownerId"    => Cerbos::Sequel.field("owner_id"),
    #     "request.resource.attr.department" => Cerbos::Sequel.field("owner.department"),
    #     "request.resource.attr.tags"       => Cerbos::Sequel.association(
    #       :tags, member_field: "name", fields: {"name" => Cerbos::Sequel.field("name")}
    #     )
    #   }
    #
    # If the attribute map does not contain a plan variable, the adapter raises
    # {UnmappedAttributeError}. It also raises that error if the operator around the variable
    # cannot use the mapping. The adapter never selects a column by itself.
    module AttributeMapping
      # How the caller sends a NULL column to Cerbos. See
      # {Cerbos::Sequel.field} and {Cerbos::Sequel.query_plan_to_dataset}.
      NULL_REPRESENTATIONS = %i[explicit omitted].freeze

      # A scalar mapping.
      #
      # +path+ is a column on the model. It can also be a path with dots through +many_to_one+
      # or +one_to_one+ associations, for example <tt>"owner.department"</tt>. The adapter
      # translates such a path into a correlated scalar subquery. Thus the path cannot
      # increase the number of rows in the result.
      #
      # +null_representation+ declares how the caller sends THIS column when it is NULL. It
      # overrides the +null_attribute_representation+ of the call for this attribute only.
      # A mapping that declares nothing keeps the value of the call.
      Field = Struct.new(:path, :null_representation) do
        def initialize(path:, null_representation: nil)
          raise ArgumentError, "path is required" if path.nil?

          unless null_representation.nil? ||
              NULL_REPRESENTATIONS.include?(null_representation.to_sym)
            raise ArgumentError,
              "null_representation must be :explicit or :omitted, got " \
              "#{null_representation.inspect}"
          end

          super(path: path.to_s, null_representation: null_representation&.to_sym)
        end

        # @return [Array<String>] the path in parts: the association hops and then the column.
        def segments
          @segments ||= path.split(".").freeze
        end
      end

      # A collection mapping.
      #
      # +association+ is the name of a +one_to_many+ or +many_to_many+ association on the model
      # that owns it. The adapter opens a +many_to_many+ into its join table and its target, in
      # one correlated subquery.
      #
      # +member_field+ replaces the element when the policy uses the collection as a list of
      # simple values. Thus <tt>"urgent" in R.attr.tags</tt> compares with <tt>tag.name</tt>
      # and not with the +Tag+ record.
      #
      # +fields+ maps the member names in the bodies of lambdas. For example,
      # <tt>R.attr.tags.exists(t, t.name == "x")</tt> needs <tt>"name"</tt>. An entry in
      # +fields+ can be an association. This is how the adapter resolves a chain with more
      # than one hop, for example
      # <tt>R.attr.categories.exists(c, c.subCategories.exists(s, ...))</tt>.
      Association = Struct.new(:association, :member_field, :fields) do
        def initialize(association:, member_field: nil, fields: {})
          raise ArgumentError, "association is required" if association.nil?

          fields.each do |name, mapping|
            unless mapping.is_a?(Field) || mapping.is_a?(Association)
              raise ArgumentError, "fields[#{name.inspect}] must be a field or association mapping"
            end
          end

          super(
            association: association.to_sym,
            member_field: member_field&.to_s,
            fields: fields.transform_keys(&:to_s).freeze
          )
        end
      end
    end

    # Makes a scalar {AttributeMapping::Field} mapping.
    #
    # @param path [String, Symbol] a column name, or a path with dots through to-one
    #   associations
    # @param null_representation [Symbol, nil] +:explicit+ if the caller sends this column as
    #   an attribute whose value is null when the column is NULL, +:omitted+ if the caller
    #   sends no attribute at all then. The default, +nil+, keeps the value that the call
    #   gives. Declare it to make +eq+, +ne+ and +in+ against this column agree with the PDP
    #   for the rows where it is NULL (cerbos/query-plan-adapters#308).
    # @return [AttributeMapping::Field]
    def self.field(path, null_representation: nil)
      AttributeMapping::Field.new(path: path, null_representation: null_representation)
    end

    # Makes a collection {AttributeMapping::Association} mapping.
    #
    # @param association [String, Symbol] the association name on the model that owns it
    # @param member_field [String, Symbol, nil] the member column that replaces a simple
    #   element value
    # @param fields [Hash{String => AttributeMapping::Field, AttributeMapping::Association}]
    #   the member names that the bodies of lambdas use
    # @return [AttributeMapping::Association]
    def self.association(association, member_field: nil, fields: {})
      AttributeMapping::Association.new(
        association: association, member_field: member_field, fields: fields
      )
    end
  end
end
